From f00f2cc5451ba39d86cf7afff0b7ac7b5a297a91 Mon Sep 17 00:00:00 2001
From: Eberhard Beilharz <eb1@sil.org>
Date: Tue, 17 Jan 2012 17:18:35 +0100
Subject: [PATCH] Xamarin-2938: Fix Screen.FromControl

We have to convert the location of the control to screen coordinates
so that we end up with the correct screen in a multi-monitor setup.
This patch fixes Screen.FromControl, Screen.FromHandle, and
Screen.GetWorkingArea.

Change-Id: I6c4fd65833d012e832d97d610c79abc792e57adf
---
 .../System.Windows.Forms/Screen.cs                 |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Screen.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Screen.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/Screen.cs
@@ -122,7 +122,8 @@
 
 		#region Public Static Methods
 		public static Screen FromControl(Control control) {
-			return Screen.FromPoint(control.Location);
+			var point = control.Parent != null ? control.Parent.PointToScreen(control.Location) : control.Location;
+			return Screen.FromPoint(point);
 		}
 
 		public static Screen FromHandle(IntPtr hwnd) {
@@ -130,7 +131,8 @@
 
 			control = Control.FromHandle(hwnd);
 			if (control != null) {
-				return Screen.FromPoint(control.Location);
+				var point = control.Parent != null ? control.Parent.PointToScreen(control.Location) : control.Location;
+				return Screen.FromPoint(point);
 			}
 			return Screen.PrimaryScreen;
 		}
